cli/sql.go

```diff
@@ -0,0 +1,167 @@
+// Copyright 2015 The Cockroach Authors.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or
+// implied. See the License for the specific language governing
+// permissions and limitations under the License. See the AUTHORS file
+// for names of contributors.
+//
+// Author: Marc berhault (marc@cockroachlabs.com)
+
+package cli
+
+import (
+	"database/sql"
+	"fmt"
+	"io"
+	"os"
+	"strings"
+	"text/tabwriter"
+
+	// Import cockroach driver.
+	_ "github.com/cockroachdb/cockroach/sql/driver"
+	"github.com/cockroachdb/cockroach/util"
+	"github.com/spf13/cobra"
+	"golang.org/x/crypto/ssh/terminal"
+)
+
+func makeSQLClient() *sql.DB {
+	// TODO(pmattis): Initialize the user to something more
+	// reasonable. Perhaps Context.Addr should be considered a URL.
+	db, err := sql.Open("cockroach",
+		fmt.Sprintf("%s://root@%s?certs=%s",
+			Context.RequestScheme(),
+			Context.Addr,
+			Context.Certs))
+	if err != nil {
+		fmt.Fprintf(osStderr, "failed to initialize SQL client: %s\n", err)
+		osExit(1)
+	}
+	return db
+}
+
+// sqlShellCmd opens a sql shell.
+var sqlShellCmd = &cobra.Command{
+	Use:   "sql [options]",
+	Short: "open a sql shell",
+	Long: `
+Open a sql shell running against the cockroach database at --addr.
+`,
+	Run: runTerm,
+}
+
+// processOneLine takes a line from the terminal,
+// determines what type of statement it is, runs it,
+// and displays the result.
+// TODO(marc): handle multi-line, this will require ';' terminated statements.
+func processOneLine(db *sql.DB, line string, w io.Writer) (bool, error) {
+	// Look for the first word in the statement and use it to decide what to do.
+	words := strings.Split(line, " ")
+	if len(words) == 0 {
+		return false, nil
+	}
+
+	// This is a really terrible way of figuring out how to handle
+	// the input.
+	command := strings.ToUpper(words[0])
+	switch command {
+	case "EXIT", "QUIT":
+		return true, nil
+	case "SELECT", "SHOW":
+		// Issues a query and examine returned Rows.
+		rows, err := db.Query(line)
```

Comment:
Btw, there is no real harm in calling `db.Query` for all SQL statements. The same code is called in the driver for `db.Query` and `db.Exec`, the only difference is the return type (the rows or a count of the rows).